### PR TITLE
Storm updates for functions and packages and some utilities

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/ast.py [hunk 1]
```diff
@@ -2716,3 +2784,103 @@ def prepare(self):
...
+
+        valu = None
+        async for node, path in genr:
+            if self.kids:
+                valu = await self.kids[0].compute(path)
+
+            raise StormReturn(valu)
+
+        # no items in pipeline... execute
+        if len(self.kids):
```
Comment (jnwatson) [inline synapse/lib/ast.py]: `if self.kids` for consistency with 5 lines above, or use a tmp variable.

File: synapse/lib/stormtypes.py [hunk 2]
```diff
@@ -899,34 +952,34 @@ def addLibFuncs(self):
...
 
         return ret
 
-    async def _libVarsSet(self, name, valu, strip=False):
+    async def _libVarsSet(self, name, valu):
         '''
         Set a variable in a storm query
         '''
-        if strip:
+        if name.startswith('$'):
```
Comment (jnwatson) [inline synapse/lib/stormtypes.py]: Why are we turning `$$$$$foo` into `foo`?

Comment (vEpiphyte) [reply synapse/lib/stormtypes.py]: ``$$$$$$$$$$$$$foo` isn't a valid variable by syntax anyways

Comment (invisig0th) [reply synapse/lib/stormtypes.py]: Going to kill off all this strip nonsense :)

File: synapse/lib/ast.py [hunk 3]
```diff
@@ -424,7 +439,12 @@ def getRuntVars(self, runt):
 
         async for node, path in genr:
 
-            async for item in s_coro.agen(await self.kids[1].compute(path)):
+            # TODO: remove when storm is all objects
+            valu = await self.kids[1].compute(path)
+            if isinstance(valu, dict):
+                valu = valu.items()
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: this is dangerous since you can raise an exception inside of iteration

File: synapse/lib/ast.py [hunk 4]
```diff
@@ -459,7 +479,12 @@ def getRuntVars(self, runt):
         # no nodes and a runt safe value should execute once
         if node is None and self.kids[1].isRuntSafe(runt):
 
-            async for item in s_coro.agen(await self.kids[1].runtval(runt)):
+            # TODO: remove when storm is all objects
+            valu = await self.kids[1].compute(runt)
+            if isinstance(valu, dict):
+                valu = valu.items()
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: this is dangerous since you can raise an exception inside of iteration

File: synapse/lib/ast.py [hunk 5]
```diff
@@ -2758,3 +2840,103 @@ def prepare(self):
...
+
+    async def callfunc(self, runt, args, kwargs):
+        '''
+        Execute a function call using the given runtime.
+
+        This function may return a value / generator / async generator
+        '''
+        argdefs = self.kids[1].value()
+        if len(args) != len(argdefs):
+            raise Exception('BAD CALL ARG COUNT FIXME')
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: fix me!

File: synapse/cortex.py [hunk 6]
```diff
@@ -977,12 +1001,115 @@ def getCmdBrief():
...
+            raise s_exc.BadPkgDef(mesg=mesg)
+
+        mods = pkgdef.get('modules', ())
+        cmds = pkgdef.get('commands', ())
+
+        for mdef in mods:
+
+            modname = mdef.get('name')
+            if modname is None:
+                raise Foo()
```
Comment (vEpiphyte) [inline synapse/cortex.py]: this needs to be a real exception

File: synapse/lib/node.py [hunk 7]
```diff
@@ -793,6 +795,34 @@ def fork(self, node):
...
+
+    def initframe(self, initvars=None, initrunt=None):
+
+        # full copy for now...
+        framevars = self.vars.copy()
+        if initvars is not None:
+            framevars.update(initvars)
+
+        if initrunt is None:
+            initrunt = self.runt
```
Comment (vEpiphyte) [inline synapse/lib/node.py]: this line is untested but also a weird semantic to test

File: synapse/lib/ast.py [hunk 8]
```diff
@@ -373,14 +388,15 @@ class InitBlock(AstNode):
 
             if not once:
                 async for innr in subq.run(runt, agen()):
-                    pass  # pragma: no cover
+                    yield innr
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: this is untested

File: synapse/lib/ast.py [hunk 9]
```diff
@@ -397,7 +413,7 @@ class FiniBlock(AstNode):
             yield item
 
         async for innr in subq.run(runt, agen()):
-            pass  # pragma: no cover
+            yield innr
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: this is untested

File: synapse/lib/ast.py [hunk 10]
```diff
@@ -424,7 +440,15 @@ def getRuntVars(self, runt):
...
         async for node, path in genr:
 
-            async for item in s_coro.agen(await self.kids[1].compute(path)):
+            # TODO: remove when storm is all objects
+            valu = await self.kids[1].compute(path)
+            if isinstance(valu, dict):
+                valu = list(valu.items())
+
+            if valu is None:
+                valu = ()
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: this is untested

File: synapse/cortex.py [hunk 11]
```diff
@@ -668,6 +668,20 @@ def setFeedOffs(self, iden, offs):
         await self._reqUserAllowed(('model', 'tagprop', 'del'))
         return await self.cell.delTagProp(name)
 
+    async def addStormPkg(self, pkgdef):
```
Comment (vEpiphyte) [inline synapse/cortex.py]: all need tests

File: synapse/cortex.py [hunk 12]
```diff
@@ -977,12 +1001,115 @@ def getCmdBrief():
...
+        name = pkgdef.get('name')
+        await self.pkghive.set(name, pkgdef)
+
+    async def delStormPkg(self, name):
+        '''
+        Delete a storm package by name.
+        '''
+        pkgdef = await self.pkghive.pop(name, None)
+        if pkgdef is None:
+            mesg = f'No storm package: {name}.'
```
Comment (vEpiphyte) [inline synapse/cortex.py]: need sad path test here

File: synapse/lib/stormtypes.py [hunk 13]
```diff
@@ -88,6 +93,30 @@ def addLibFuncs(self):
...
+    def addLibFuncs(self):
+        self.locls.update({
+            'add': self._libPkgAdd,
+            #'get': self._libPkgGet,
+            'del': self._libPkgDel,
+            'list': self._libPkgList,
+        })
+
+    async def _libPkgAdd(self, pkgdef):
+        self.runt.reqAllowed(('storm', 'pkg', 'add'))
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py]: needs test

File: synapse/lib/ast.py [hunk 14]
```diff
@@ -2758,3 +2849,104 @@ def prepare(self):
...
+
+            try:
+
+                async for item in self.kids[2].run(funcrunt, agen()):
+                    pass
+
+            except StormReturn as e:
+                return e.item
+
+            return None
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: should this raise instead of return?

File: synapse/lib/ast.py [hunk 15]
```diff
@@ -2758,3 +2849,104 @@ def prepare(self):
...
+        runt.setVar(name, realfunc)
+
+        async for node, path in genr:
+            path.setVar(name, realfunc)
+            yield node, path
+
+    def getRuntVars(self, runt):
+        yield self.kids[0].value()
+
+    async def callfunc(self, runt, args, kwargs):
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: need a kwargs example / use plumbed?

File: synapse/lib/ast.py [hunk 16]
```diff
@@ -2758,3 +2849,124 @@ def prepare(self):
...
+
+    async def run(self, runt, genr):
+
+        # fake out a generator...
+        for item in ():
+            yield item
+
+        valu = None
+        async for node, path in genr:
+            if self.kids:
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: untested bit

File: synapse/cortex.py [hunk 17]
```diff
@@ -981,12 +1007,115 @@ def getCmdBrief():
...
+        return self.stormpkgs.get(name)
+
+    async def getStormPkgs(self):
+        return list(self.pkghive.values())
+
+    async def getStormMods(self):
+        return self.stormmods
+
+    async def getStormMod(self, name):
+        return self.stormmods.get(name)
```
Comment (vEpiphyte) [inline synapse/cortex.py]: potentially vestigial

File: synapse/lib/ast.py [hunk 18]
```diff
@@ -710,19 +785,31 @@ class LiftOper(Oper):
...
+        if node is None and self.kids[0].isRuntSafe(runt):
+            valu = await self.kids[0].compute(runt)
+            async for subn in self.yieldFromValu(runt, valu):
+                yield subn, runt.initPath(subn)
 
     async def yieldFromValu(self, runt, valu):
 
+        # there is nothing in None... ;)
+        if valu is None:
+            return
```
Comment (vEpiphyte) [inline synapse/lib/ast.py]: untested

**Pull request merged**